risc0/circuit/bigint/src/zkr.rs

```diff
@@ -0,0 +1,131 @@
+// Copyright 2024 RISC Zero, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::BIGINT_PO2;
+use anyhow::{ensure, Context, Result};
+use risc0_circuit_recursion::prove::Program;
+use risc0_core::field::baby_bear::BabyBearElem;
+use risc0_zkp::adapter::TapsProvider;
+use std::io::Read;
+
+pub(crate) const ZKR_ZIP: &[u8] = include_bytes!("bigint_zkr.zip");
+
+fn get_zkr_u32s(name: &str) -> Result<Vec<u32>> {
+    let mut zip = zip::ZipArchive::new(std::io::Cursor::new(ZKR_ZIP))?;
+    let mut f = zip
+        .by_name(name)
+        .with_context(|| format!("Failed to read {name}"))?;
+
+    let mut u8vec: Vec<u8> = Vec::new();
+    f.read_to_end(&mut u8vec)?;
+
+    Ok(Vec::from(bytemuck::cast_slice(u8vec.as_slice())))
+}
+
+pub fn get_zkr(name: &str, po2: usize) -> Result<Program> {
+    let code_size = risc0_circuit_recursion::CIRCUIT
+        .get_taps()
+        .group_size(risc0_circuit_recursion::REGISTER_GROUP_CODE);
+
+    let u32s = get_zkr_u32s(name)?;
+    ensure!(u32s.len() % code_size == 0);
+    ensure!(u32s.len() / code_size < (1 << po2));
+    let rows = u32s.len() / code_size;
+    let pct = (rows as f64 * 100.) / ((1 << po2) as f64);
+    tracing::debug!(
+        "Got {}/{} ({:.2}%) rows of program for {name} with po2={po2}",
+        rows,
+        1 << po2,
+        pct
+    );
+
+    Ok(Program {
+        code: u32s.iter().cloned().map(BabyBearElem::from).collect(),
+        code_size,
+        po2,
+    })
+}
+
+static REGISTER_ZKRS: std::sync::Once = std::sync::Once::new();
+
+pub fn register_zkrs() {
```

Comment:
When/how does this get called?

Reply:
https://github.com/risc0/risc0/blob/50e505c828d41391fe654c31da753452d73aeec2/risc0/circuit/bigint/src/rsa_tests.rs#L300